clang-tidy-checks/potential-unbounded-loop-check.diff

```diff
@@ -0,0 +1,3051 @@
++using namespace clang::ast_matchers;
++
++namespace clang {
++namespace tidy {
++namespace misc {
++
++void PotentialUnboundedLoopCheck::registerMatchers(MatchFinder *Finder) {
++  Finder->addMatcher(
++      forStmt(allOf(hasCondition(
++                        binaryOperator(
++                            hasLHS(hasDescendant(declRefExpr().bind("lhs"))),
++                            hasRHS(hasDescendant(declRefExpr().bind("rhs"))),
++                            anyOf(hasOperatorName("<"), hasOperatorName("<="),
++                                  hasOperatorName(">"), hasOperatorName(">=")))
++                            .bind("op")),
++                    hasIncrement(
++                        unaryOperator(
++                            anyOf(hasOperatorName("++"), hasOperatorName("--")),
++                            hasDescendant(declRefExpr().bind("inc")))
++                            .bind("inc_op"))))
++          .bind("for"),
++      this);
++}
++
++inline bool isInc(UnaryOperatorKind Op) {
++  return Op == UnaryOperatorKind::UO_PostInc ||
++         Op == UnaryOperatorKind::UO_PreInc;
++}
++
++inline bool isDec(UnaryOperatorKind Op) {
++  return Op == UnaryOperatorKind::UO_PostDec ||
++         Op == UnaryOperatorKind::UO_PreDec;
++}
++
++inline bool isLT(BinaryOperatorKind Op) {
++  return Op == BinaryOperatorKind::BO_LT;
++}
++
++inline bool isLE(BinaryOperatorKind Op) {
++  return Op == BinaryOperatorKind::BO_LE;
++}
++inline bool isGT(BinaryOperatorKind Op) {
++  return Op == BinaryOperatorKind::BO_GT;
++}
++
++inline bool isGE(BinaryOperatorKind Op) {
++  return Op == BinaryOperatorKind::BO_GE;
++}
++
++void PotentialUnboundedLoopCheck::check(
++    const MatchFinder::MatchResult &Result) {
++  const auto *LHSDecl = Result.Nodes.getNodeAs<DeclRefExpr>("lhs")->getDecl();
++  const auto *RHSDecl = Result.Nodes.getNodeAs<DeclRefExpr>("rhs")->getDecl();
++  const auto *IncDecl = Result.Nodes.getNodeAs<DeclRefExpr>("inc")->getDecl();
++  auto Op = Result.Nodes.getNodeAs<BinaryOperator>("op")->getOpcode();
++  auto IncOp = Result.Nodes.getNodeAs<UnaryOperator>("inc_op")->getOpcode();
++  const auto *MatchedFor = Result.Nodes.getNodeAs<ForStmt>("for");
++
++  // only analyze non-compound types
++  if (LHSDecl->getType()->isCompoundType() ||
++      RHSDecl->getType()->isCompoundType() ||
++      IncDecl->getType()->isCompoundType()) {
++    return;
++  }
++
++  // only analyze loops in which increment var used in condition
++  if (IncDecl->getName() != LHSDecl->getName() &&
++      IncDecl->getName() != RHSDecl->getName()) {
++    return;
++  }
++
++  auto LHSWidth =
++      LHSDecl->getASTContext().getTypeInfo(LHSDecl->getType()).Width;
++  auto RHSWidth =
++      RHSDecl->getASTContext().getTypeInfo(RHSDecl->getType()).Width;
++
++  if (LHSWidth < RHSWidth) {
++    if (((isLT(Op) || isLE(Op)) && isInc(IncOp)) ||
++        ((isGT(Op) || isGE(Op)) && isDec(IncOp))) {
++      diag(MatchedFor->getForLoc(), "Potential unbounded loop");
```

Comment:
Could this be a little more verbose (preferably with some data)? Something like `Found LHS with size X and RHS with size Y, this can potentially lead to an unbound loop. Explicitly cast the RHS before comparing to fix this warning.`